devito/finite_differences/derivative.py

```diff
@@ -555,12 +615,23 @@ def _eval_expand_add(self, **hints):
         `Derivative(f(x) + g(x), x)
             --> Derivative(f(x), x) + Derivative(g(x), x)`
         '''
-        expr = self.args[0]
-        if isinstance(expr, sympy.Add):
-            ind, dep = expr.as_independent(*self.dims, as_Add=True)
-            if isinstance(dep, sympy.Add):
+        if self.expr.is_Add:
+            ind, dep = self.expr.as_independent(*self.dims, as_Add=True)
+            if dep.is_Add:
                 return Add(*[self.func(s, *self.args[1:]) for s in dep.args])
             else:
                 return self.func(dep, *self.args[1:])
         else:
             return self
+
+    def _eval_expand_product_rule(self, **hints):
+        ''' Expands products, of functions of the dependent variable
+        `Derivative(f(x)·g(x), x)
+            --> Derivative(f(x), x)·g(x) + f(x)·Derivative(g(x), x)`
+        This is only implemented for first derivatives with an arbitrary number
+        of multiplicands and second derivatives with two multiplicands. The
+        resultant expression for higher derivatives and mixed derivatives is much
+        more difficult to implement.
```

Comment:
Guessing it needs some variant on https://en.wikipedia.org/wiki/Fa%C3%A0_di_Bruno%27s_formula ?